File: ncurses/scoreboard.cpp
#include "scoreboard.h"
#include "legend.h"
#include "visettings.h"

Scoreboard::Scoreboard( int x, int y, int width, int height ) : Window( x, y, width, height )
{

}

void Scoreboard::update()
{
  wattrset( mWindow, COLOR_PAIR( INSTRUCTIONS ) );

  int line = 1;

  int red = 0;
  int blue = 0;

  for( std::vector<Plant>::iterator i = getAttr( state )->plants.begin(); i != getAttr( state )->plants.end(); i++ )
  {
    // Count Red Plants
    if( i->ownerID == 0 )
      ++red;
    else
      ++blue;
  }

  setAttr( player1Plants, red );
  setAttr( player2Plants, blue );

  mvwprintw( mWindow, line++, 2, "Turn: %d      ", getAttr( turnNumber ) );

  wattrset( mWindow, COLOR_PAIR( PLAYER_1_COLOR ) );
  mvwprintw( mWindow, line++, 2, "---------------------------" );
  wattron( mWindow, A_BOLD );
  mvwprintw( mWindow, line++, 3, getAttr( player1Name ) );
  wattroff( mWindow, A_BOLD );
  mvwprintw( mWindow, line++, 2, "---------------------------" );
  mvwprintw( mWindow, line++, 3, "Score: %d       ", getAttr( player1Score ) );
  mvwprintw( mWindow, line++, 3, "Light: %d       ", getAttr( player1Light ) );
  mvwprintw( mWindow, line++, 3, "Plants: %d      ", red );

  int highestLevel[2] = {0,0};
  for( vector<Plant>::iterator i = getAttr( state )->plants.begin(); i != getAttr( state )->plants.end(); i++ )
  {
    if( i->rootLevel + i->leafLevel + i->flowerLevel > highestLevel[i->ownerID] )
    {
      highestLevel[i->ownerID] = i->rootLevel + i->leafLevel + i->flowerLevel;
    }
  }

  mvwprintw( mWindow, line++, 3, "Plant Level: %d      ", highestLevel[0] );

  wattrset( mWindow, COLOR_PAIR( PLAYER_2_COLOR ) );

  line = 9;

  mvwprintw( mWindow, line++, 2, "---------------------------" );
  wattron( mWindow, A_BOLD );
  mvwprintw( mWindow, line++, 3, getAttr( player2Name ) );
  wattroff( mWindow, A_BOLD );
  mvwprintw( mWindow, line++, 2, "---------------------------" );
  mvwprintw( mWindow, line++, 3, "Score: %d       ", getAttr( player2Score ) );
  mvwprintw( mWindow, line++, 3, "Light: %d       ", getAttr( player2Light ) );
  mvwprintw( mWindow, line++, 3, "Plants: %d      ", blue );
  mvwprintw( mWindow, line++, 3, "Plant level: %d     ", highestLevel[1] );

  wattrset( mWindow, COLOR_PAIR( INSTRUCTIONS ) );

  line = getAttr( maxY )*2 - 12;

  mvwprintw( mWindow, line++, 3, "Visualizer Usage:" );
  mvwprintw( mWindow, line++, 3, "Q     | Quit" );
  mvwprintw( mWindow, line++, 3, "P     | Pause/Play" );
  mvwprintw( mWindow, line++, 3, "R     | Rewind" );
  mvwprintw( mWindow, line++, 3, "+     | Increase Speed" );
  mvwprintw( mWindow, line++, 3, "-     | Decrease Speed" );
  mvwprintw( mWindow, line++, 3, ">     | Go ahead one frame" );
  mvwprintw( mWindow, line++, 3, "<     | Go back one frame" );
  mvwprintw( mWindow, line++, 3, "E     | Go to last frame" );
  mvwprintw( mWindow, line++, 3, "S     | Go to first frame" );
  mvwprintw( mWindow, line++, 3, "Left  | Move Cursor Left" );
  mvwprintw( mWindow, line++, 3, "Right | Move Cursor Right" );
  mvwprintw( mWindow, line++, 3, "Up    | Move Cursor Up" );
  mvwprintw( mWindow, line++, 3, "Down  | Move Cursor Down" );


  wattroff( mWindow, COLOR_PAIR( 8 ) );

  wrefresh( mWindow );

}

File: pycurses/scoreboard.py
# -*- coding: utf-8 -*-
"""
Created on Mon Apr 12 00:49:41 2010

@author: jacob
"""

import curses
from legend import Legend
from window import Window
from visettings import Visettings

        
class Scoreboard( Window ):
    def __init__(self, x, y, width, height):
        Window.__init__( self, x, y, width, height) 

    def printLine( self, n, str ):
        self.Window.addstr( n, 3, str )
        n+=1
        return n

    def update( self ):
        Window.update( self )
        
        line = 1
        
        red = 0
        blue = 0
        
        highestLevel = [0,0]
        for i in Visettings.State.plants:
            if i.ownerID: 
                blue += 1
            else:
                red += 1
                
            if i.rootLevel + i.leafLevel + i.flowerLevel > highestLevel[i.ownerID]:
                highestLevel[i.ownerID] = i.rootLevel + i.leafLevel + i.flowerLevel 
                
        self.Window.addstr( line, 2, "Turn: %d     " % (Visettings.FrameNumber) )
        line+=1
        
        self.Window.attrset( curses.color_pair( Legend.PLAYER_1_COLOR  ) )
        self.Window.addstr( line, 2, "----------------------------" )
        line+=1
        line = self.printLine( line, Visettings.Player1Name )
        self.Window.addstr( line, 2, "----------------------------" )
        line+=1
        line = self.printLine( line, "Score: %d     " % (Visettings.Player1Score) )
        line = self.printLine( line, "Light: %d     " % (Visettings.Player1Light) )
        line = self.printLine( line, "Plants: %d     " % (red) )
        
        self.Window.attrset( curses.color_pair( Legend.PLAYER_2_COLOR  ) )
        line+=2
        self.Window.addstr( line, 2, "----------------------------" )
        line+=1
        line = self.printLine( line, Visettings.Player2Name )
        self.Window.addstr( line, 2, "----------------------------" )
        line+=1
        line = self.printLine( line, "Score: %d     " % (Visettings.Player2Score) )
        line = self.printLine( line, "Light: %d     " % (Visettings.Player2Light) )
        line = self.printLine( line, "Plants: %d     " % (blue) )
        line = self.printLine( line, "Plant Level: %d     " % (highestLevel[1]) )
        
        self.Window.attrset( curses.color_pair( Legend.INSTRUCTIONS ) )
        
        line = Visettings.MaxY*2-12
        
        Visettings.Player1Plants = red
        Visettings.Player2Plants = blue
        
        instructions = [
        "Visualizer Usage: ",
        "Q     | Quit",
        "P     | Pause/Play",
        "R     | Rewind",
        "+     | Increase Speed",
        "-     | Decrease Speed",
        ">     | Go ahead one frame",
        "<     | Go back one frame",
        "E     | Go to last frame ",
        "S     | Go to first frame",
        "Left  | Move Cursor Left",
        "Right | Move Cursor Right",
        "Up    | Move Cursor Up",
        "Down  | Move Cursor Down"]
        
        for k in instructions:
            line = self.printLine( line, k )
        
        self.Window.refresh()
        Visettings.Screen.refresh()
